clippy_lints/src/methods/unnecessary_join.rs

```diff
@@ -0,0 +1,45 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::ty::is_type_diagnostic_item;
+use hir::ExprKind;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+use rustc_span::sym;
+
+use super::UNNECESSARY_JOIN;
+
+pub(super) fn check<'tcx>(context: &LateContext<'tcx>, join: &'tcx hir::Expr<'tcx>) {
+    let applicability = Applicability::MachineApplicable;
+
+    if_chain! {
+        if let ExprKind::MethodCall(_path, [join_self_arg, join_arg], _span) = &join.kind;
+        let collect_output_type = context.typeck_results().expr_ty(join_self_arg);
+        // the current join method is being called on a vector
+        // e.g .join("")
+        if is_type_diagnostic_item(context, collect_output_type, sym::Vec);
```

Comment:
I think this check is redundant. I would just move the check for `::<Vec<String>>` up and remove this check.

Reply:
Fair enough, will do

Reply:
Done